### PR TITLE
feat(pubsublite): default routing policy

### PR DESCRIPTION
**Pull request by 18suresha**
This PR implements the `DefaultRoutingPolicy` for a multipartition publisher.

This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553)

### CODE REVIEW

File: google/cloud/pubsublite/internal/default_routing_policy_test.cc [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+namespace {
+
+using google::cloud::pubsublite_internal::DefaultRoutingPolicy;
+
+TEST(DefaultRoutingPolicyTest, RouteWithKey) {
+  std::string data = R"""({
+  "oaisdhfoiahsd": 18,
```
Comment (dpcollins-google) [inline google/cloud/pubsublite/internal/default_routing_policy_test.cc:29]: If you're not loading the JSON file, instead just use std::unordered_map<std::string, int> directly

**Review by coryan — CHANGES_REQUESTED** [review]
Please split this into two PRs, one refactoring the SHA256 code and one adding the default routing policy.

On the PR adding the default routing policy, please add something explaining why you chose that particular algorithm.  I trust there are good reasons for it, I would like to know them.

> Reviewed 1 of 2 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-:-Mz0lx5P7xwpI4pGuJKO:b-lmj804)* status: 4 of 26 files reviewed, 11 unresolved discussions (waiting on @18suresha, @coryan, and @dpcollins-google)

**Review by 18suresha — COMMENTED** [review]
https://github.com/googleapis/google-cloud-cpp/pull/8597

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-:-Mz1-v3L4foQssqiDlZN:b6ljwo5)* status: 4 of 26 files reviewed, 11 unresolved discussions (waiting on @18suresha, @coryan, and @dpcollins-google)

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r10, 1 of 3 files at r17, 4 of 13 files at r19, 1 of 2 files at r20, 1 of 1 files at r21, 17 of 17 files at r23, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553)* status: all files reviewed, 18 unresolved discussions (waiting on @18suresha)

___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 44 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29HGH6YOxmSQlbS-s:-Mz29HGH6YOxmSQlbS-t:b1eadkr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L44)):*
> ```cpp
>     std::uint64_t pow_mod = ModPow(
>         // 2^8
>         static_cast<std::uint64_t>(2 * 2 * 2 * 2 * 2 * 2 * 2 * 2), pow, mod);
> ```

I would have written that as `1 << 8`, but then again: I think generally I could use some comments explaining what motivates the `GetMod()` and `ModPow()` functions.  Why do we want them?
___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 52 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29bvJ5Ba77kYdUnUr:-Mz29bvJ5Ba77kYdUnUs:b-n8q52e) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L52)):*
> ```cpp
> }
> 
> std::uint64_t DefaultRoutingPolicy::RouteWithoutKey(
> ```

This is a case where overloading seems sensible?  `Route(std::uint32_t num_partitions)` is obviously without a key?
___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 54 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29l2t-KqXiKqXXFt4:-Mz29l2t-KqXiKqXXFt5:b-ovi6rp) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L54)):*
> ```cpp
> std::uint64_t DefaultRoutingPolicy::RouteWithoutKey(
>     std::uint32_t num_partitions) {
>   return counter_.fetch_add(1, std::memory_order_relaxed) % num_partitions;
> ```

Do we really think the relaxed memory order will make a difference?  Or put another way: do we have any evidence that it will make any difference?
___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 59 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz2E-kQ4V4uzILht37w:-Mz2E-kQ4V4uzILht37x:b-6ao5gr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L59)):*
> ```cpp
> std::uint64_t DefaultRoutingPolicy::Route(std::string message_key,
>                                           std::uint32_t num_partitions) {
>   return GetMod(google::cloud::internal::Sha256Hash(std::move(message_key)),
> ```

I see.... Is there a reason why we cannot use good hash function here instead of doing the Sha256Hash and this special mod?

Ah, after reading the test I think I understand. We need to use a hash function and modulo that matches the other client libraries.  All the more reason to document that in the class definition.  Otherwise somebody (me from the future for example) could just "simplify" the code and since we only test with C++ I won't even notice that I broken interoperability.
___
*[google/cloud/pubsublite/internal/routing_policy.h, line 25 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz2EYCuE2H-g76AZtis:-Mz2EYCuE2H-g76AZtit:b-j49ymz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/routing_policy.h#L25)):*
> ```objc
> namespace pubsublite_internal {
> 
> class RoutingPolicy {
> ```

Please add some description of what this interface is for.
___
*[google/cloud/pubsublite/internal/routing_policy.h, line 27 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz2EbL_2wGxB0BBnc6e:-Mz2EbL_2wGxB0BBnc6f:bu2e5tk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/routing_policy.h#L27)):*
> ```objc
> class RoutingPolicy {
>  public:
>   virtual std::uint64_t RouteWithoutKey(std::uint32_t num_partitions) = 0;
> ```

add `virtual ~RoutingPolicy() = default`.
___
*[google/cloud/pubsublite/internal/routing_policy.h, line 28 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz2EN5g9Rnci7uVN3gY:-Mz2EN5g9Rnci7uVN3gZ:b-qjf0zz) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/routing_policy.h#L28)):*
> ```objc
>  public:
>   virtual std::uint64_t RouteWithoutKey(std::uint32_t num_partitions) = 0;
>   virtual std::uint64_t Route(std::string message_key,
> ```

Seems unlikely that the routing policy will consume the `message_key`?  Should this be `std::string const& message_key`?

**Review by 18suresha — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553)* status: all files reviewed, 18 unresolved discussions (waiting on @18suresha and @coryan)

___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 44 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29HGH6YOxmSQlbS-s:-Mz2PgDc9Y5W5_-ctihs:b51kggp) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L44)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I would have written that as `1 << 8`, but then again: I think generally I could use some comments explaining what motivates the `GetMod()` and `ModPow()` functions.  Why do we want them?
</blockquote></details>

added comments
___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 52 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29bvJ5Ba77kYdUnUr:-Mz2PhgN14_12c4j9Xk0:b-ru80en) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L52)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

This is a case where overloading seems sensible?  `Route(std::uint32_t num_partitions)` is obviously without a key?
</blockquote></details>

yup, changed
___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 54 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29l2t-KqXiKqXXFt4:-Mz2PkqYC0CfV05E0tLo:b-9e0f8d) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L54)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Do we really think the relaxed memory order will make a difference?  Or put another way: do we have any evidence that it will make any difference?
</blockquote></details>

'Typical use for relaxed memory ordering is incrementing counters' from https://en.cppreference.com/w/cpp/atomic/memory_order
___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 59 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz2E-kQ4V4uzILht37w:-Mz2Q8Zh3DQPdMYmkG1f:b51kggp) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L59)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I see.... Is there a reason why we cannot use good hash function here instead of doing the Sha256Hash and this special mod?

Ah, after reading the test I think I understand. We need to use a hash function and modulo that matches the other client libraries.  All the more reason to document that in the class definition.  Otherwise somebody (me from the future for example) could just "simplify" the code and since we only test with C++ I won't even notice that I broken interoperability.

</blockquote></details>

added comments

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r24, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553)* status: all files reviewed, 14 unresolved discussions (waiting on @18suresha)

___
*[google/cloud/pubsublite/internal/default_routing_policy.h, line 43 at r24](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz5Vx9T5BhowhzwClkB:-Mz5Vx9T5BhowhzwClkC:b-3f25ms) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a7acc51c1c0698689477663bbd7bf0a19031b885/google/cloud/pubsublite/internal/default_routing_policy.h#L43)):*
> ```objc
> 
>   // Uses SHA-256 as it's provided in the standard library of all silver
>   // languages, enabling consistent hashing across languages.
> ```

nit: this seems like it should be the description of the class, consider something like:


```cc
/**
  * Implements the same routing policy as all the other Pub/Sub Lite client libraries.
  *
  * All the client libraries provided by Google use the same algorithm to route messages.
  *
... insert a description of the algorithm here, and a link to the documentation if there is any ...
  */
class DefaultRoutingPolicy ...
```
___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 54 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29l2t-KqXiKqXXFt4:-Mz5U_PR6G1kyPkxJQeX:biv98uk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L54)):*
<details><summary><i>Previously, 18suresha (Adarsh Suresh) wrote…</i></summary><blockquote>

'Typical use for relaxed memory ordering is incrementing counters' from https://en.cppreference.com/w/cpp/atomic/memory_order
</blockquote></details>

That does not answer the question. `++counter_ % num_partitions` is more readable, does not make me wonder if there is some magical reason why we want to relax the memory order, and unless you have some reason to believe otherwise, it is unlikely to be measurably slower.  The question was whether you had a reason to think if it was slower.  If not, then readability probably wins.
___
*[google/cloud/pubsublite/internal/routing_policy.h, line 35 at r24](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz5V_gr3GKxhZls8LMA:-Mz5V_gr3GKxhZls8LMB:b6xmyyk) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a7acc51c1c0698689477663bbd7bf0a19031b885/google/cloud/pubsublite/internal/routing_policy.h#L35)):*
> ```objc
>   virtual std::uint64_t Route(std::string const& message_key,
>                               std::uint32_t num_partitions) = 0;
>   virtual ~RoutingPolicy() = default;
> ```

nit: please move the destructor to the beginning of the class.

**Review by 18suresha — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553)* status: 23 of 26 files reviewed, 14 unresolved discussions (waiting on @18suresha and @coryan)

___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 54 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29l2t-KqXiKqXXFt4:-Mz6USQyChvlAuSB9b4t:bhx3y59) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L54)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

That does not answer the question. `++counter_ % num_partitions` is more readable, does not make me wonder if there is some magical reason why we want to relax the memory order, and unless you have some reason to believe otherwise, it is unlikely to be measurably slower.  The question was whether you had a reason to think if it was slower.  If not, then readability probably wins.
</blockquote></details>

spoke to Dan and he said removing `std::memory_order_relaxed` is fine.

**Review by devbww — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553)* status: 23 of 26 files reviewed, 14 unresolved discussions (waiting on @18suresha and @coryan)

___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 54 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29l2t-KqXiKqXXFt4:-Mz7D-3ZEppuIyO5sEEG:b-ecw3re) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L54)):*
> `fetch_add` is a post-increment expression though so I kept it as `return counter_.fetch_add(1) % num_partitions;`

It seems to me that the readability question is independent of any pre/post-increment question.  `std::atomic<Integral>` supports both forms.

**Review by 18suresha — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553)* status: 23 of 26 files reviewed, 14 unresolved discussions (waiting on @18suresha and @coryan)

___
*[google/cloud/pubsublite/internal/default_routing_policy.cc, line 54 at r23](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-Mz29l2t-KqXiKqXXFt4:-Mz7IO_b-Vt1OfckL74A:b-cs3mj8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a70426e80cd02471d6a1282732ef9dcc9868ee37/google/cloud/pubsublite/internal/default_routing_policy.cc#L54)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

> `fetch_add` is a post-increment expression though so I kept it as `return counter_.fetch_add(1) % num_partitions;`

It seems to me that the readability question is independent of any pre/post-increment question.  `std::atomic<Integral>` supports both forms.
</blockquote></details>

oh yea I removed that part of the comment on GitHub since I realized it doesn't make sense w.r.t to Carlos' last comment, but I guess the change doesn't reflect in Reviewable
we preferred the use of `fetch_add` though it's a soft preference, so changed it to `counter_++`

**Review by coryan — APPROVED** [review]
A few nits that I trust you will fix before merging. 

> Reviewed 2 of 2 files at r25, 1 of 1 files at r28, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-:-MzFtY0m8oJJoAMDtHex:bqfrkg0)* status: all files reviewed, 13 unresolved discussions (waiting on @18suresha and @devbww)

___
*[google/cloud/pubsublite/internal/default_routing_policy.h, line 45 at r28](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-MzFsb1tC7J95Zb1gyPC:-MzFsb1uEYYlje7Uo43g:b-wh0uju) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/fc04088d4384fa917c200667dcf507d63437827f/google/cloud/pubsublite/internal/default_routing_policy.h#L45)):*
> ```objc
>  * messages.
>  *
>  * @note Algorithm for routing with a message key is <big-endian integer
> ```

nit: "The algorithm" 
___
*[google/cloud/pubsublite/internal/default_routing_policy.h, line 47 at r28](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8553#-MzFsuqmC0l_lNR_y9F8:-MzFsuqmC0l_lNR_y9F9:bh2qr98) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/fc04088d4384fa917c200667dcf507d63437827f/google/cloud/pubsublite/internal/default_routing_policy.h#L47)):*
> ```objc
>  * @note Algorithm for routing with a message key is <big-endian integer
>  * representation of SHA256(message key)> % <number of partitions>. Uses SHA-256
>  * as it's provided in the standard library of all silver languages, enabling
> ```

I do not think we speak of silver languages, specially not in the context of C++.  I would say something like "It uses SHA-256 as it is available in most programming languages."

**Pull request merged**